docs/manpages/replfwd.1.md
= REPLFWD(1)
Open-Xchange 2019
doctype: manpage

## NAME
replfwd - daemon to forward replication events between wforce clusters

## SYNOPSIS
replfwd [*OPTION*]... 

## DESCRIPTION
**replfwd** is a daemon that will forward replication events between
wforce clusters and other replication forwarders, and vice-versa.

**replfwd** configuration uses Lua, see the replfwd.conf file for a
simple example.

## SCOPE
**replfwd** is not standalone - it is dependent on a wforce cluster to
forward received events from other replication forwarders, and on
other replication forwarders to send events to.

**replfwd** would typically be configured with all the other members
  of the local wforce cluster, as well as the replication forwarders
  for all remote clusters.

## OPTIONS
-C,--config *FILE*
:    Load configuration from *FILE*.

-s
:    Run in foreground, and notify systemd when started. Use this switch to run
     replfwd inside a supervisor (use with e.g. systemd and daemontools).

-D,--docker
:    Enable timestamps in log messages for running in docker containers.

-l,--loglevel
:    Log level as an integer. 0 is Emerg, 7 is Debug. Defaults to 6 (Info). Only applies
     to stdout logging.

-d,--daemon
:    Operate as a daemon.

-f,--facility *FACILITY NAME*
:    Log using the specified facility name, e.g. local0

-k,--key
:    Generate a key to be used as the encryption key for messages,
     print to stdout and exit. Note that this is not typically
     expected to be used.

-h,--help
:    Display a helpful message and exit.

## BUGS

## SEE ALSO
replfwd.conf(5)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).

## [3.0.0]

### Added
- Support new `fail_type` parameter for determining why a login failed
- New livez and readyz endpoints (unauthenticated) for k8s environments
- Support JA3 allow/block lists in API/Lua
- Allow ja3 to be passed to the reset API command
- Add support for building amazon-2023 packages
- Use asciidoctor to build documentation not pandoc
- Add a sample grafana dashboard for monitoring wforce
- Implemented the (existing but previously always 0) `wforce_active_http_connections` metric for counting active HTTP connections

### Removed
- Removed support for Enterprise Linux 7 and Amazon 2

## [2.12.1]

### Added
- Set default map size for sharded Stats DBs based on the number of shards
- Change debian postinst scripts to modify 'setKey' and 'webpwd' parameter only if no weakforce is already installed
- Add tests for crypto code

## [2.12.0]

### Added
- Now builds a separate luajit package, based on the openresty luajit fork. This is to address some issues found with stock luajit. The package also includes some lua modules that wforce typically makes use of.

### Changed
- Build the wforce-minimal image for both arm64 and amd64, and add provenance.
- Fix centos-7/el-7 builds to still work after centos-7 went EOL
- Add support for debian-bookworm, remove support for debian-buster

## [2.10.2]

### Changed
- Fixed LuaState selection algorithm to use a free pool, which should lead to faster/more 
  consistent selection of lua states by threads.

### Added
- CI now builds an additional image 'powerdns/wforce-minimal' using alpine for more secure and much smaller image

## [2.10.1]

### Changed
- Fixed bug in GeoIP2 lookups where return values were not populated

## [2.10.0]

### Added
- Add Enterprise Linux 9 Build Target
- Option to use OpenSSL instead of Libsodium for encryption

### Removed
- Remove Legacy GeoIP from Packages and Dockerfiles/Images
- Remove the report_api from weakforced entirely

### Changed
- Move to pytest instead of nose for regression tests

## [2.8.0]

### Added 
- Support ELK 7.x Stack 
- Support Date Expansion in WebHook URLs
- Enable IP and Login substitution in blocklist return messages
- Add config option to disable password for /metrics endpoint
- Support redis usernames and passwords for redis authentication
- Support hostnames for redis configuration in addition to IP addresses

## Changed

- Fix an issue where IPv6 ComboAddress returned zero port number (which caused v6 HTTP listen addresses to not work)
- Set V6ONLY socket option to stop v6 sockets from managing v4 addresses for replication
- Return the IP address of the client in JSON of ACL denied response

## [2.6.1]
- Fix issue where wforce was complaining about not being able to create tmp file on startup
- Fix timing issue whereby the webserver was not started before syncDB leading to syncDone failures
- Use debian bullseye-slim in wforce docker image
- Fix issue in wforce docker image where the default config file was overridden with a volume mount but not used

## [2.6.0]
- Support HTTPS in webserver using libdrogon
- Add support for configuring TLS behaviour of outbound HTTPS connections
- Remove support for building debian stretch packages
- Change behaviour of default ACLs such that they are overridden by setACL()
- Build for debian bullseye

## [2.4.1]

### Added
- New API and Lua commands to dynamically manage siblings
- Siblings can now have unique encryption keys
- Amazon Linux packaging target for pdns-builder

### Changed
- Fix issue where replication length bytes can be truncated causing syncDB problems

## [2.4.0]

### Added
- New wf_dump_entries tool to dump stats DBs to file
- Support for new "forwarding" type in replication messages
- Support for Prometheus via the new /metrics REST endpoint
- Enable TCP keepalive for redis connections
- Configurable timeout for R/W on redis connections

### Changed
- Fix duplicate command stats under some circumstances

## [2.2.2]

### Added
- Support building debian buster packages
- StatsDBs can now be sharded for better performance
- StatsDB expiry thread now runs more often by default
- StatsDB expiry thread sleep time is now configurable
- Support for Kafka REST Proxy in webhooks

### Changed
- Fix control socket leak when client closes connection immediately

## [2.2.1]

### Changed
- Fix wforce crash in Sibling send thread triggered by syncDB operation 

## [2.2.0]

### Added
- Lookup individual GeoIP2 values (Unsigned Integer, String,
Double/Float and Boolean)
- Add session_id to wforce LoginTuple
- The report_api is now deployable via proper packaging, systemd and gunicorn
- Custom GET Endpoints supporting non-JSON return values
- New Kibana Reports and Dashboard to view effectiveness of wforce
policy
- New "type" field added to built-in webhooks
- Built-in whitelists added in addition to built-in blacklists
- New checkXXXBlacklist (and Whitelist) function in Lua
- Built-in black/whitelisting can be disabled and checked from Lua
instead
- Thread names support
- Blacklisting return messages are configurable
- Add support for TCP keepalive
- New functions to retrieve the custom return message for blacklisted IPs, Logins and IP/Logins.

### Changed
- Fix typo in wforce.conf.example blackistLogin->blacklistLogin
- Python3 support throughout wforce (regression tests and deployment)
- Uses "bytes" instead of "string" in replication, which avoids
protobuf errors for non-UTF-8 StatsDB keys or values.
- Remove ctpl from wforce
- Sibling threads now use explicit std::queue instead of ctpl
- Perform replication to siblings asynchronously instead of synchronously,
  preventing delays at startup and in REST API functions which triggered
  replication.

## [2.0.0]

### Added
- Add configuration setting "setNumWebHookConnsPerThread"

wforce/replfwd.cc
/*
 * This file is part of PowerDNS or weakforced.
 * Copyright -- PowerDNS.COM B.V. and its contributors
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of version 3 of the GNU General Public License as
 * published by the Free Software Foundation.
 *
 * In addition, for the avoidance of any doubt, permission is granted to
 * link this program with OpenSSL and to (re)distribute the binaries
 * produced as the result of such linking.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include "config.h"
#include <stddef.h>
#define SYSLOG_NAMES
#include <syslog.h>
#include "sstuff.hh"
#include "misc.hh"
#include <netinet/tcp.h>
#include <limits>
#include <thread>
#include "dolog.hh"
#include <readline/readline.h>
#include <readline/history.h>
#include "base64.hh"
#include <fstream>
#include <unistd.h>
#include <sys/stat.h>
#include "sodcrypto.hh"
#include "lock.hh"
#include "replfwd.hh"
#include "replfwd-web.hh"
#include "replfwd-replication.hh"
#include "wforce-prometheus.hh"
#include "iputils.hh"
#include "ext/threadname.hh"

#include <getopt.h>
#ifdef HAVE_LIBSYSTEMD
#include <systemd/sd-daemon.h>
#endif
#include "device_parser.hh"
#include "wforce_ns.hh"
#include "wforce-replication.hh"

using std::atomic;
using std::thread;
bool g_verbose=false;
bool g_console=true;
bool g_docker=false;

struct WForceStats g_stats;
WforceWebserver g_webserver;
WebHookRunner g_webhook_runner;
WebHookDB g_webhook_db;
string g_outputBuffer;
LogLevel g_loglevel{LogLevel::Info};

std::string g_configDir; // where the config files are located
ReplfwdReplication g_replication; //  handling replication

struct
{
  bool beDaemon{false};
  bool underSystemd{false};
  bool beClient{false};
  string command;
  string config;
  string regexes;
  unsigned int facility{LOG_DAEMON};
} g_cmdLine;

std::mutex g_luamutex;
LuaContext g_lua;

void replicateOperation(const ReplicationOperation& rep_op)
{
  g_replication.replicateOperation(rep_op);
}

static void daemonize(void)
{
  if(fork())
    _exit(0); // bye bye
  
  setsid(); 

  int i=open("/dev/null",O_RDWR); /* open stdin */
  if(i < 0) 
    ; // L<<Logger::Critical<<"Unable to open /dev/null: "<<stringerror()<<endl;
  else {
    dup2(i,0); /* stdin */
    dup2(i,1); /* stderr */
    dup2(i,2); /* stderr */
    close(i);
  }
}

std::string findDefaultConfigFile()
{
  std::string configFile = string(SYSCONFDIR) + "/wforce/replfwd.conf";
  struct stat statbuf;

  if (stat(configFile.c_str(), &statbuf) != 0) {
    g_configDir = string(SYSCONFDIR);
    configFile = string(SYSCONFDIR) + "/replfwd.conf";
  }
  else
    g_configDir = string(SYSCONFDIR) + "/wforce";
  return configFile;
}

int main(int argc, char** argv)
try
{
  g_stats.latency=0;

  signal(SIGPIPE, SIG_IGN);
  signal(SIGCHLD, SIG_IGN);

#ifdef HAVE_LIBSODIUM
  if (sodium_init() == -1) {
    cerr<<"Unable to initialize crypto library"<<endl;
    exit(EXIT_FAILURE);
  }
#endif
  g_cmdLine.config = findDefaultConfigFile();
  struct option longopts[]={ 
    {"config", required_argument, 0, 'C'},
    {"systemd",  optional_argument, 0, 's'},
    {"daemon", optional_argument, 0, 'd'},
    {"docker", optional_argument, 0, 'D'},
    {"facility", required_argument, 0, 'f'},
    {"key", optional_argument, 0, 'k'},
    {"loglevel", required_argument, 0, 'l'},
    {"help", 0, 0, 'h'}, 
    {0,0,0,0} 
  };
  int longindex=0;
  for(;;) {
    int c=getopt_long(argc, argv, ":hsdDkC:f:l:v", longopts, &longindex);
    if(c==-1)
      break;
    switch(c) {
    case 'C':
      g_cmdLine.config=getFileFromPath(optarg);
      g_configDir = getDirectoryPath(optarg);
      break;
    case 'd':
      g_console = false;
      g_cmdLine.beDaemon=true;
      break;
    case 's':
      g_console = false;
      g_cmdLine.underSystemd=true;
      break;
    case 'D':
      g_docker = true;
      break;
    case 'f':
      int i;
      for (i = 0; facilitynames[i].c_name; i++)
        if (strcmp((char *)facilitynames[i].c_name, optarg)==0)
          break;
      if (facilitynames[i].c_name) {
        g_cmdLine.facility = facilitynames[i].c_val;
      }
      else {
        cout << "Bad log facility" << endl;
        exit(1);
        break;
      }
      break;
    case 'l':
      try {
        g_loglevel = static_cast<LogLevel>(std::stoi(optarg));
      }
      catch (const std::invalid_argument &ia) {
        cout << "Bad log level (" << optarg << ") - must be an integer" << endl;
        exit(1);
      }
      break;
    case 'k':
      {
        std::string newkey{newKey()};
        cout << newkey << endl;
        exit(EXIT_SUCCESS);
        break;
      }
    case 'h':
      cout<<"Syntax: replfwd [-C,--config file] [-D, --docker] [-d,--daemon] [-f,--facility-name facility] [-l --loglevel 0-7] [-k,--key] [-h,--help]\n";
      cout<<"\n";
      cout<<"-C,--config file      Load configuration from 'file'\n";
      cout<<"-s,                   Operate under systemd control.\n";
      cout<<"-d,--daemon           Operate as a daemon\n";
      cout<<"-D,--docker           Enable timestamp logging for docker\n";
      cout<<"-f,--facility name    Use log facility 'name'\n";
      cout<<"-l,--loglevel level   Log level as an integer. 0 is Emerg, 7 is Debug. Defaults to 6 (Info).\n";
      cout<<"-k,--key              Create and output a key, then exit\n";
      cout<<"-h,--help             Display this helpful message\n";
      cout<<"\n";
      exit(EXIT_SUCCESS);
      break;
    case 'v':
      g_verbose=true;
      break;
    case '?':
    default:
      cout << "Option '-" << (char)optopt << "' is invalid: ignored\n";
    }
  }
  argc-=optind;
  argv+=optind;

  g_webserver.setWebLogLevel(g_loglevel);
  openlog("replfwd", LOG_PID, g_cmdLine.facility);
  
  g_singleThreaded = false;
  if (chdir(g_configDir.c_str()) != 0) {
    warnlog("Could not change working directory to %s (%s)", g_configDir, strerror(errno));
  }

  // Initialise Prometheus Metrics
  initWforcePrometheusMetrics(std::make_shared<WforcePrometheus>("replfwd"));

  // Setup a sensible ACL, but allow it to be overridden by Lua if necessary
  auto acl = g_webserver.getACL();
  for(auto& addr : {"127.0.0.0/8", "10.0.0.0/8", "100.64.0.0/10", "169.254.0.0/16", "192.168.0.0/16", "172.16.0.0/12", "::1/128", "fc00::/7", "fe80::/10"})
    acl.addMask(addr);
  g_webserver.setACL(acl);

  // this sets up the global lua state used for config and setup
  auto todo=setupLua(g_lua, g_cmdLine.config);

  if(g_cmdLine.beDaemon) {
    daemonize();
  }
  else {
    vinfolog("Running in the foreground");
  }

  // register all the webserver commands
  registerWebserverCommands();

  vector<string> vec;
  std::string acls;
  acl.toStringVector(&vec);
  for(const auto& s : vec) {
    if (!acls.empty())
      acls += ", ";
    acls += s;
  }
  noticelog("ACL allowing queries from: %s", acls.c_str());

  // Start the replication worker threads
  g_replication.startReplicationWorkerThreads();

  // start the threads created by lua setup. Includes the webserver accept thread
  for(auto& t : todo)
    t();

#ifdef HAVE_LIBSYSTEMD
  sd_notify(0, "READY=1");
#endif

  while (true)
    pause();
  _exit(EXIT_SUCCESS);

 }
catch(const LuaContext::ExecutionErrorException& e) {
  try {
    errlog("Fatal Lua error: %s", e.what());
    std::rethrow_if_nested(e);
  }
  catch(const std::exception& e) {
    errlog("Details: %s", e.what());
  }
  catch(WforceException &ae)
    {
      errlog("Fatal replfwd error: %s", ae.reason);
    }
  _exit(EXIT_FAILURE);
 }
 catch(std::exception &e) {
   errlog("Fatal error: %s", e.what());
 }
 catch(WforceException &ae) {
   errlog("Fatal replfwd error: %s", ae.reason);
   _exit(EXIT_FAILURE);
 }
